tartufo/scanner.py

```diff
@@ -247,13 +248,38 @@ def included_paths(self) -> List[Pattern]:
         """
         if self._included_paths is None:
             self.logger.info("Initializing included paths")
-            patterns = list(self.global_options.include_path_patterns or ())
-            self._included_paths = (
-                config.compile_path_rules(set(patterns)) if patterns else []
-            )
-            self.logger.debug(
-                "Included paths was initialized as: %s", self._included_paths
-            )
+            patterns = list(self.global_options.include_path_patterns or ())  # type: ignore
+            try:
+                patterns = [
+                    pattern["path-pattern"]
+                    for pattern in cast(List[Dict[str, str]], patterns)
+                ]
+                self._included_paths = (
+                    config.compile_path_rules(set(cast(List[str], patterns)))
+                    if patterns
+                    else []
+                )
+                return self._included_paths
```

Comment:
I believe we need this return as we need to stop further execution if the user has new style of include and exclude path patterns

Reply:
So, a couple of things. We're stuck with butt-ugly code because of the logic, but I think a couple of tactical maneuvers could help us out. These comments are inclusive for the block from L259 through L298.

First, we have a problem in that although we are telling people they must use "all old flavor" or "all new flavor", the code actually enforces "all processed toml files must use the same old or new flavor" which isn't what we say, may not be practicable, and isn't necessary.

To combat this, I think we need to process the `patterns` (global) and `include_patterns` (local) lists independently, and combine them at the end after they have been compiled. Tactically, I would also suggest avoiding reuse of `patterns` when we know the type is changing and use of `set` rather than `list` for the result so we don't have to deduplicate manually (and subsequent queries should be more efficient).

In pseudocode, something more like:

```
self._included_paths: Set[Pattern] = set()
for plist in (self.global_options.include_path_patterns, self.config_data.get("include_path_patterns")):
    if plist:
        try:
            self._included_paths.update(
                config.compile_path_rules({x["path-patterns"] for x in plist})
            )
            continue # New way worked, move on
        except TypeError:
            warnings.warn(blah blah blah, DeprecationWarning)
        # Try the old way
        try:
            self._included_paths.update(
                config.compile_path_rules(set(plist))
            )
        except TypeError as exc:
             raise types.ConfigException(you can't mix both types in the same list)
```
We possibly do some extra regex compiling before deduplicating, but I think the clarity is worth the overhead.

Reply:
...and I'd use the same pattern (excuse the pun) for the exclude processing that comes next.

Reply:
...and yes, this changes the return from `List[Pattern]` to `Set[Pattern]` but I can't see anything in the code base cares, and I would expect set membership testing to be more efficient than list membership testing.